src/Math/Lexer.php

```diff
@@ -0,0 +1,319 @@
+use Drupal\rules\Math\Exception\UnknownTokenException;
+use Drupal\rules\Math\Token\ParenthesesCloseToken;
+use Drupal\rules\Math\Token\ParenthesesOpenToken;
+use Drupal\rules\Math\Token\CommaToken;
+use Drupal\rules\Math\Token\ConstantToken;
+use Drupal\rules\Math\Token\FunctionToken;
+use Drupal\rules\Math\Token\NumberToken;
+use Drupal\rules\Math\Token\OperatorTokenInterface;
+use Symfony\Component\Process\Exception\LogicException;
+
+class Lexer {
+
+  /**
+   * Static cache of the compiled regular expression.
+   *
+   * @var string
+   */
+  protected $compiledRegex;
+
+  /**
+   * The list of registered operators.
+   *
+   * @var string[]
+   */
+  protected $operators = [];
+
+  /**
+   * The list of registered constants.
+   *
+   * @var number[]
+   */
+  protected $constants = [];
+
+  /**
+   * The list of registered functions.
+   *
+   * @var array[]
+   */
+  protected $functions = [];
+
+  /**
+   * Registers a function with the lexer.
+   *
+   * @param string $name
+   *   The name of the function.
+   * @param callable $function
+   *   The callback to be executed.
+   * @param int $arguments
+   *   The number of arguments required for the function.
+   *
+   * @return $this
+   */
+  public function addFunction($name, callable $function, $arguments = 1) {
+    $this->functions[$name] = [$arguments, $function];
+    return $this;
+  }
+
+  /**
+   * Registers an operator with the lexer.
+   *
+   * @param string $regex
+   *   The regular expression of the operator token.
+   * @param string $operator
+   *   The full qualified class name of the operator token.
+   *
+   * @return $this
+   */
+  public function addOperator($regex, $operator) {
+    if (!in_array('Drupal\rules\Math\Token\OperatorTokenInterface', class_implements($operator))) {
+      throw new LogicException();
+    }
+
+    // Clear the static cache when a new operator is added.
+    unset($this->compiledRegex);
+
+    $this->operators[md5($regex)] = [$regex, $operator];
+    return $this;
+  }
+
+  /**
+   * Registers a constant with the lexer.
+   *
+   * @param string $name
+   *   The name of the constant.
+   * @param int $value
+   *   The value of the constant.
+   *
+   * @return $this
+   */
+  public function addConstant($name, $value) {
+    $this->constants[$name] = $value;
+    return $this;
+  }
+
+  /**
+   * Generates a token stream from a mathematical expression.
+   *
+   * @param string $input
+   *   The mathematical expression to tokenize.
+   *
+   * @return array
+   *   The generated token stream.
+   */
+  public function tokenize($input) {
+    $matches = [];
+    $regex = $this->getCompiledTokenRegex();
+
+    if (preg_match_all($regex, $input, $matches, PREG_SET_ORDER | PREG_OFFSET_CAPTURE) === FALSE) {
+      throw new \LogicException();
+    };
+
+    $types = [
+      'number', 'operator', 'function', 'open', 'close', 'comma', 'constant',
+    ];
+
+    return array_map(function ($match) use ($types) {
```

Comment:
Looks like beautiful magic to me. I think it would be great to describe whats the data structure should look like. As we don't have interfaces describing stuff, maybe the methods should be a bit more verbose on what they are doing